### ISSUE TITLE
Incorrect IDE0004 (Cast is redundant) in sign extended operations

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#20617]
**Version Used**: 15.3 Preview 3

:link: Originally filed internally as [bug 398604](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?id=398604).

**Steps to Reproduce**:

Add the following method:

```csharp
public static void MethodName()
{
    const long x = ~(long)~1U;
    switch (x)
    {
    case 1:
        var y = x;

    default:
        break;
    }
}
```

**Expected Behavior**:

IDE0004 is not reported, and no code fix tries to remove the cast to `long`.

**Actual Behavior**:

IDE0004 is reported. When the code fix removes the cast to `long`, the program no longer compiles.

**Comment by sharwell** [comment]
Here's a unit test volunteered by @jcouv:

```VB
        <Fact>
        <WorkItem(20617, "https://github.com/dotnet/roslyn/issues/20617")>
        Public Async Function TestCSharp_Remove_Bug398604() As Task
            Dim input =
<Workspace>
    <Project Language="C#" CommonReferences="true">
        <Document>
class C
{
    void M()
    {
        long x = ~{|Simplify:(long)|}~(uint)1;
    }
}
        </Document>
    </Project>
</Workspace>

            Dim expected =
<code>
class C
{
    void M()
    {
        long x = ~~(uint)1;
    }
}
</code>

            Await TestAsync(input, expected)

        End Function
```
